### PR TITLE
Bert4rec, make it use sequence dataset with last iten prediction sampler, add loss batch_logsoftmax

### PR DESCRIPTION
**Pull request by YaroslavZhurba**

### CODE REVIEW

File: modeling/models/bert4rec.py [hunk 1]
```diff
@@ -10,6 +10,7 @@ def __init__(
             self,
             sequence_prefix,
             labels_prefix,
+            candidate_prefix,
```
Comment (YaroslavZhurba) [inline modeling/models/bert4rec.py]: я давно не делал pull master и так много всего изменилось, вернул для совместимости

File: modeling/models/bert4rec.py [hunk 2]
```diff
@@ -48,11 +50,14 @@ def __init__(
 
         self._init_weights(initializer_range)
 
+        self._cls_token = nn.Parameter(torch.rand(1, 1, embedding_dim))
```
Comment (YaroslavZhurba) [inline modeling/models/bert4rec.py]: здесь заменил на rand, по идее сейчас создается вектор от 0 до 1 с рандомными значениями

File: modeling/models/bert4rec.py [hunk 3]
```diff
@@ -67,31 +72,40 @@ def forward(self, inputs):
...
-            candidate_scores = self._get_last_embedding(embeddings, mask)  # (batch_size, num_items)
+            candidate_scores = torch.einsum(
+                'bd,nd->bn',
+                predictions,
+                self._item_embeddings.weight
+            )  # (batch_size, num_items + 2)
             candidate_scores[:, 0] = -torch.inf
             candidate_scores[:, self._num_items + 1:] = -torch.inf
 
+            if '{}.ids'.format(self._candidate_prefix) in inputs: # only validation should be here
```
Comment (YaroslavZhurba) [inline modeling/models/bert4rec.py]: это похоже, сейчас вообще работать не будет. Я обновил свой код сделава пул мастера и смерджив его и почему-то массово пропали все валидационные семплеры, не понял что произошло, код пока оставил

File: configs/train/bert4rec_train_config.json [hunk 4]
```diff
@@ -2,14 +2,13 @@
   "experiment_name": "bert4rec_beauty",
   "best_metric": "eval/ndcg@20",
   "dataset": {
-    "type": "scientific",
```
Comment (NonameUntitled) [inline configs/train/bert4rec_train_config.json]: Конфиги надо не менять, а сделать новые с пометкой, что это bert4rec_new или что-то в этом роде

File: modeling/loss/base.py [hunk 5]
```diff
@@ -53,6 +54,38 @@ def forward(self, inputs):
...
+class BatchLogSoftmaxLoss(TorchLoss, config_name='batch_logsoftmax'):
+
+    def __init__(self, predictions_prefix, candidates_prefix):
+        super().__init__()
+        self._predictions_prefix = predictions_prefix
+        self._candidates_prefix = candidates_prefix
+
+    @classmethod
+    def create_from_config(cls, config, **kwargs):
+        return cls(predictions_prefix=config.get('predictions_prefix'), candidates_prefix=config.get('candidates_prefix'))
```
Comment (NonameUntitled) [inline modeling/loss/base.py]: Тут надо разнести по разным строчкам `predictions_prefix` и `candidates_prefix`, чтобы проще читалось

File: modeling/loss/base.py [hunk 6]
```diff
@@ -53,6 +54,38 @@ def forward(self, inputs):
...
+    def forward(self, inputs): # use log soft max
+        predictions = inputs[self._predictions_prefix]
+        candidates = inputs[self._candidates_prefix]
+
+        # Step 1: Compute the dot product between each pair of vectors from output and candidates
+        # Result will be of shape (batch_size, batch_size)
+        dot_product_matrix = predictions @ candidates.T
+
+        # Step 2: Apply softmax to each row of the resulting tensor
+        softmax_matrix = F.softmax(dot_product_matrix, dim=1)
```
Comment (NonameUntitled) [inline modeling/loss/base.py]: Ты говорил, что эту часть нормально перепишешь с использованием [вот этого](https://pytorch.org/docs/stable/generated/torch.nn.LogSoftmax.html)

File: modeling/models/base.py [hunk 7]
```diff
@@ -91,7 +91,7 @@ def __init__(
         )
         self._encoder = nn.TransformerEncoder(transformer_encoder_layer, num_layers)
 
-    def _apply_sequential_encoder(self, events, lengths):
+    def _apply_sequential_encoder(self, events, lengths, add_cls_token = False):
```
Comment (NonameUntitled) [inline modeling/models/base.py]:
```suggestion
    def _apply_sequential_encoder(self, events, lengths, add_cls_token=False):
```

File: modeling/models/base.py [hunk 8]
```diff
@@ -122,6 +122,12 @@ def _apply_sequential_encoder(self, events, lengths):
 
         embeddings[~mask] = 0
 
+        if add_cls_token:
+            cls_token_expanded = self._cls_token.expand(batch_size, -1, -1) # !!! use torch tile
+            embeddings = torch.cat((cls_token_expanded, embeddings), dim=1)
+            mask = torch.cat((torch.ones((batch_size, 1), dtype=torch.bool, device=DEVICE), mask), dim=1)
+
```
Comment (NonameUntitled) [inline modeling/models/base.py:130]: Лишняя пустая строка

File: modeling/models/base.py [hunk 9]
```diff
@@ -122,6 +122,12 @@ def _apply_sequential_encoder(self, events, lengths):
 
         embeddings[~mask] = 0
 
+        if add_cls_token:
+            cls_token_expanded = self._cls_token.expand(batch_size, -1, -1) # !!! use torch tile
```
Comment (NonameUntitled) [inline modeling/models/base.py]: А комментарий `# !!! use torch tile` это просто шутка?)

File: modeling/models/bert4rec.py (hunk 2 above)
Comment (NonameUntitled) [reply modeling/models/bert4rec.py]: Потом не забудь заменить на нормальную размерность

File: modeling/models/bert4rec.py (hunk 1 above)
Comment (NonameUntitled) [reply modeling/models/bert4rec.py]: Вообще я подумал, вроде бы я это реально выкручивал, так что можно убрать кандидатов и все, что с ними связяно

File: modeling/models/bert4rec.py [hunk 10]
```diff
@@ -67,31 +72,40 @@ def forward(self, inputs):
         all_sample_events = inputs['{}.ids'.format(self._sequence_prefix)]  # (all_batch_events)
         all_sample_lengths = inputs['{}.length'.format(self._sequence_prefix)]  # (batch_size)
 
+        add_cls_token = True
```
Comment (NonameUntitled) [inline modeling/models/bert4rec.py]: Зачем ты выделяешь в переменную то, что используется только один раз?

File: modeling/models/bert4rec.py
Comment (NonameUntitled) [inline modeling/models/bert4rec.py:1]: Та же ошибка, что и выше, надо не переписывать bert4rec, а сделать новый условно bet4rec_new

File: modeling/models/bert4rec.py (hunk 3 above)
Comment (NonameUntitled) [reply modeling/models/bert4rec.py]: Да, я их убирал, так и должно быть, потом расскажу почему, а сейчас просто пойми, что валидация не нужна пока

**Review by NonameUntitled — CHANGES_REQUESTED** [review]

File: modeling/models/bert4rec.py (hunk 2 above)
Comment (YaroslavZhurba) [reply modeling/models/bert4rec.py]: окей, буду использовать squeeze, unsqueeze

File: modeling/models/bert4rec.py (hunk 3 above)
Comment (YaroslavZhurba) [reply modeling/models/bert4rec.py]: разобрались, ненужно и убираем префиксы

File: configs/train/bert4rec_train_config.json (hunk 4 above)
Comment (YaroslavZhurba) [reply configs/train/bert4rec_train_config.json]: окей, сделаю

File: modeling/loss/base.py (hunk 5 above)
Comment (YaroslavZhurba) [reply modeling/loss/base.py]: тут ясно

File: modeling/loss/base.py (hunk 6 above)
Comment (YaroslavZhurba) [reply modeling/loss/base.py]: окей, справлю

File: modeling/models/base.py (hunk 7 above)
Comment (YaroslavZhurba) [reply modeling/models/base.py]: pep8, проверить

File: .gitignore [hunk 11]
```diff
@@ -1,2 +1,9 @@
 .idea
 __pycache__
+data/Beauty/test.pkl
```
Comment (NonameUntitled) [inline .gitignore]: Можешь написать просто `data/*`

File: modeling/models/bert4rec_cls.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+import torch.nn as nn
+
+
+class Bert4RecModelCLS(SequentialTorchModel, config_name='bert4rec_cls'):
+
+    def __init__(
+            self,
+            sequence_prefix,
+            labels_prefix,
+            candidate_prefix,
```
Comment (NonameUntitled) [inline modeling/models/bert4rec_cls.py]: Мы же вчера обсуждали, что это больше не нужно

File: modeling/models/bert4rec_cls.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+            num_layers=num_layers,
+            dim_feedforward=dim_feedforward,
+            dropout=dropout,
+            activation=activation,
+            layer_norm_eps=layer_norm_eps,
+            is_causal=False
+        )
+        self._sequence_prefix = sequence_prefix
+        self._labels_prefix = labels_prefix
+        self._candidate_prefix = candidate_prefix
```
Comment (NonameUntitled) [inline modeling/models/bert4rec_cls.py]: И это не нужно

File: modeling/models/bert4rec_cls.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        self._init_weights(initializer_range)
+
+        self._cls_token = nn.Parameter(torch.rand(embedding_dim))
+
+    @classmethod
+    def create_from_config(cls, config, **kwargs):
+        return cls(
+            sequence_prefix=config['sequence_prefix'],
+            labels_prefix=config['labels_prefix'],
+            candidate_prefix=config['candidate_prefix'],
```
Comment (NonameUntitled) [inline modeling/models/bert4rec_cls.py]: И это не нужно

File: modeling/models/bert4rec_cls.py [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        )
+
+        self._bias = nn.Parameter(
+            data=torch.zeros(num_items + 2),
+            requires_grad=True
+        )
+
+        self._init_weights(initializer_range)
+
+        self._cls_token = nn.Parameter(torch.rand(embedding_dim))
```
Comment (NonameUntitled) [inline modeling/models/bert4rec_cls.py:49]: Это надо до `self._init_weights(initializer_range)` делать

File: modeling/models/bert4rec_cls.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+                candidate_lengths = inputs['{}.length'.format(self._candidate_prefix)]  # (batch_size)
+
+                batch_size = candidate_lengths.shape[0]
+                num_candidates = candidate_lengths[0]
+
+                candidate_scores = torch.gather(
+                    input=candidate_scores,
+                    dim=1,
+                    index=torch.reshape(candidate_events, [batch_size, num_candidates])
+                )  # (batch_size, num_candidates)
```
Comment (NonameUntitled) [inline modeling/models/bert4rec_cls.py]: И это тоже надо убрать

File: modeling/dataset/base.py
Comment (NonameUntitled) [inline modeling/dataset/base.py:1]: Отмени в этом файле все форматирования, которые ты сделал

File: modeling/loss/base.py [hunk 17]
```diff
@@ -53,6 +54,36 @@ def forward(self, inputs):
...
+            candidates_prefix=config.get('candidates_prefix')
+        )
+
+    def forward(self, inputs):  # use log soft max
+        predictions = inputs[self._predictions_prefix]
+        candidates = inputs[self._candidates_prefix]
+
+        dot_product_matrix = predictions @ candidates.T
+
+        m = nn.LogSoftmax(dim=1)
```
Comment (NonameUntitled) [inline modeling/loss/base.py]: что за `m`?

**Review by NonameUntitled — CHANGES_REQUESTED** [review]

File: modeling/dataset/base.py [hunk 18]
```diff
@@ -217,13 +217,13 @@ def create_from_config(cls, config, **kwargs):
             )
             max_user_idx, max_item_idx = max(max_user_idx, train_max_user_idx), max(max_item_idx, train_max_item_idx)
             max_sequence_length = max(max_sequence_length, train_max_sequence_length)
-            
+
             validation_dataset[domain], validation_max_user_idx, validation_max_item_idx, validation_max_sequence_length = cls._create_dataset(
```
Comment (YaroslavZhurba) [inline modeling/dataset/base.py:221]: это что-то не получается венуть, думаю, что тут не принципиально

Comment (YaroslavZhurba) [reply modeling/dataset/base.py:221]: все эти пробелы/отступы в этой файле

**Review by NonameUntitled — APPROVED** [review]
